File: wrappers/fsharp/TurboToken.fs
module TurboToken.Api

open System
open System.Runtime.InteropServices

let version () : string =
    let ptr = Native.turbotoken_version()
    if ptr = IntPtr.Zero then "unknown"
    else
        let s = Marshal.PtrToStringAnsi(ptr)
        if isNull s then "unknown" else s

let clearCache () : unit =
    Native.turbotoken_clear_rank_table_cache()

let getEncoding (name: string) : Async<Result<Encoding.Encoding, string>> =
    async {
        match Registry.getEncodingSpec name with
        | Error e -> return Error e
        | Ok spec ->
            let! rankResult = RankCache.readRankFile spec.Name
            match rankResult with
            | Error e -> return Error e
            | Ok rankData ->
                return Ok { Encoding.Name = spec.Name; Encoding.Spec = spec; Encoding.RankPayload = rankData }
    }

let getEncodingForModel (model: string) : Async<Result<Encoding.Encoding, string>> =
    async {
        match Registry.modelToEncoding model with
        | Error e -> return Error e
        | Ok encodingName -> return! getEncoding encodingName
    }

let listEncodingNames () : string list =
    Registry.listEncodingNames ()

File: wrappers/php/src/TurboToken.php
<?php

declare(strict_types=1);

namespace TurboToken;

class TurboToken
{
    /** @var array<string, Encoding> */
    private static array $cache = [];

    public static function getEncoding(string $name): Encoding
    {
        if (!isset(self::$cache[$name])) {
            $spec = Registry::getEncodingSpec($name);
            $rankPayload = RankCache::readRankFile($name);
            self::$cache[$name] = new Encoding($rankPayload, $spec);
        }
        return self::$cache[$name];
    }

    public static function getEncodingForModel(string $model): Encoding
    {
        $name = Registry::modelToEncoding($model);
        return self::getEncoding($name);
    }

    /**
     * @return string[]
     */
    public static function listEncodingNames(): array
    {
        return Registry::listEncodingNames();
    }

    public static function version(): string
    {
        return NativeBridge::version();
    }
}
